x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java

```diff
@@ -703,6 +700,29 @@ private void assertRecovered(String indexName, TotalHits originalAllHits, TotalH
         }
     }
 
+    private void assertRecoveryStats(String indexName, boolean preWarmEnabled) {
+        int shardCount = getNumShards(indexName).totalNumShards;
+        final RecoveryResponse recoveryResponse = client().admin().indices().prepareRecoveries(indexName).get();
+        assertThat(recoveryResponse.shardRecoveryStates().get(indexName).size(), equalTo(shardCount));
+
+        for (List<RecoveryState> recoveryStates : recoveryResponse.shardRecoveryStates().values()) {
+            for (RecoveryState recoveryState : recoveryStates) {
+                logger.info("Checking {}[{}]", recoveryState.getShardId(), recoveryState.getPrimary() ? "p" : "r");
+                ByteSizeValue cacheSize = getCacheSizeForShard(recoveryState.getShardId());
+                boolean largeCache = cacheSize.compareTo(new ByteSizeValue(1, ByteSizeUnit.MB)) >= 0;
+                assertThat(
+                    Strings.toString(recoveryState),
+                    recoveryState.getIndex().recoveredFileCount(),
+                    preWarmEnabled && largeCache ? greaterThan(0) : greaterThanOrEqualTo(0)
+                );
+
+                // Since the cache size is variable, the pre-warm phase might fail as some of the files can be evicted
+                // while a part is pre-fetched, in that case the recovery state is left as FINALIZE.
+                assertThat(recoveryState.getStage(), anyOf(equalTo(RecoveryState.Stage.DONE), equalTo(RecoveryState.Stage.FINALIZE)));
```

Comment:
That's not true anymore with the latest approach? Shouldn't it always be DONE?

Reply:
In those tests the cache size is set randomly, meaning that's possible that in some test instances the cache won't have room for all the files in the index therefore some of the pre-warming tasks would fail.

Reply:
Can we make a stronger assertion here then in case where the cache is set to unbounded?